src/Castle.Core/DynamicProxy/Generators/GeneratorUtil.cs

```diff
@@ -74,25 +74,23 @@ bool IsReadOnly(ParameterInfo parameter)
 				//
 				// The above points inform the following detection logic: First, we rely on an IL
 				// `[in]` modifier being present. This is a "fast guard" against non-`in` parameters:
-				if ((parameter.Attributes & (ParameterAttributes.In | ParameterAttributes.Out)) == ParameterAttributes.In)
-				{
-					// Here we perform the actual check. We don't rely on cmods because support
-					// for them is at current too unreliable in general, and because we wouldn't
-					// be saving much time anyway.
-					if (parameter.GetCustomAttributes(false).Any(IsIsReadOnlyAttribute))
-					{
-						return true;
-					}
-				}
+				if ((parameter.Attributes & (ParameterAttributes.In | ParameterAttributes.Out)) != ParameterAttributes.In)
+					return false;
 
-				return false;
+#if FEATURE_EMIT_CUSTOMMODIFIERS
+				// Check for the required modifiers first, as it's faster.
+				// This check allows to make the detection logic more robust on the platforms which support custom modifiers.
+				// If check fails, resort to the IsReadOnlyAttribute check.
+				if (parameter.GetRequiredCustomModifiers().Any(x => x == typeof(InAttribute)))
+					return true;
+#endif
 
-				bool IsIsReadOnlyAttribute(object attribute)
-				{
-					// The comparison by name is intentional; any assembly could define that attribute.
-					// See explanation in comment above.
-					return attribute.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute";
-				}
+                // The comparison by name is intentional; any assembly could define that attribute.
+                // See explanation in comment above.
+				if (parameter.GetCustomAttributes(false).Any(x => x.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute"))
+					return true;
+
+				return false;
```

Comment:
Is there a reason&mdash;beyond personal coding style preferences&mdash;for rewriting so much of this code (e.g. rewriting `IsIsReadOnlyAttribute` as an inline lambda; removing one level of indentation by inverting the first `if` condition)?

Unless I'm mistaken, this diff here could be much smaller: You could simply insert the new check (from `#if FEATURE_EMIT_CUSTOMMODIFIERS` up to and including `#endif`) above the code comment line that reads, _"Here we perform the actual check [...]"_ (which would have to be rephrased to keep it accurate), the rest could stay as it is.

Reply:
(Besides, there's some bad indentation of the code comments above the last `if`.)

Reply:
> (Besides, there's some bad indentation of the code comments above the last if.)

I'll fix it, sorry about that. Forgot to run R#'s reformat code 😊 

> Is there a reason—beyond personal coding style preferences—for rewriting so much of this code ... ?

The main reason was readability. The less level of indentation we have, the easier it's read the code. If you find it worse - I can revert.

> e.g. rewriting IsIsReadOnlyAttribute as an inline lambda;

This is optimization. With previous implementation the delegate was never cached:
![image](https://user-images.githubusercontent.com/1074182/44881809-9e538100-acb9-11e8-9b44-ef8e293bbc2f.png)

With new one it is:
![image](https://user-images.githubusercontent.com/1074182/44882088-934d2080-acba-11e8-9e97-c1f13d20b243.png)

> Unless I'm mistaken, this diff here could be much smaller

Well, my effort was to left the readable code after myself, rather then make as small diff as possible. However, if you find some changes redundant - I'm fine to refactor them, just point the direction 😉

Reply:
No, I'm fine with leaving it as you've rewritten it.